community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java

```diff
@@ -120,147 +140,90 @@ public PropertyItem get()
 
     private void readPropertyData() throws IOException
     {
-        buffer.clear();
-
         switch ( type )
         {
-            case BOOL:
-            case BYTE:
-            {
-                buffer.put( (byte) header );
-                break;
-            }
-
-            case SHORT:
-            {
-                buffer.putShort( (short) header );
-                break;
-            }
-
-            case CHAR:
-            {
-                buffer.putChar( (char) header );
-                break;
-            }
+        case STRING:
+            readFromStore( stringStore );
+            break;
+        case ARRAY:
+            readFromStore( arrayStore );
+            break;
+        case SHORT_STRING:
+        case SHORT_ARRAY:
+            ensureValuesLoaded();
+            break;
+
+        default:
+            throw new IllegalStateException();
+        }
+    }
 
-            case INT:
+    private void ensureValuesLoaded()
+    {
+        if ( remainingBlocksToRead > 0 )
+        {
+            if ( valuesAfterHeader == null || valuesAfterHeader.length < remainingBlocksToRead )
             {
-                buffer.putInt( (int) header );
-                break;
+                valuesAfterHeader = new long[remainingBlocksToRead];
             }
 
-            case LONG:
+            int offset = cursor.getOffset();
+            try
             {
-                if ( (header & 0x1L) > 0 )
-                {
-                    buffer.putLong( header >>> 1 );
-                }
-                else
+                do
                 {
-                    buffer.putLong( cursor.getLong() );
-                    remainingBlocksToRead--;
+                    cursor.setOffset( offset );
+                    for ( int i = 0; remainingBlocksToRead-- > 0; i++ )
```

Comment:
(related to previous comment, just to keep it separate) can we split loop invariant `remainingBlocksToRead-- > 0` into 2 independent statements?